src/erasure-code/clay/ErasureCodeClay.cc

```diff
@@ -0,0 +1,938 @@
+  err |= to_int("w", profile, &w, DEFAULT_W, ss);
+  
+  err |= sanity_check_k(k, ss);
+
+  err |= to_int("d", profile, &d, std::to_string(k+m-1), ss);
+
+  //check for mds block input
+  if (profile.find("scalar_mds") == profile.end() ||
+      profile.find("scalar_mds")->second.size() == 0) {
+    mds.profile["plugin"] = "jerasure";
+    pft.profile["plugin"] = "jerasure";
+    *ss << "scalar_mds not found in profile picking default jerasure" << std::endl;
+  } else {
+    std::string p = profile.find("scalar_mds")->second;
+    *ss << "recieved scalar_mds as " << p << std::endl;
+
+    if ((p == "jerasure") || (p == "isal")) {
+      mds.profile["plugin"] = p;
+      pft.profile["plugin"] = p;
+    }
+    else {
+      mds.profile["plugin"] = "jerasure";
+      pft.profile["plugin"] = "jerasure";
+      *ss << "scalar_mds " << p << "is not currently supported using the default reed_sol_van" << std::endl;
+    }
+  }
+
+  if (profile.find("technique") == profile.end() ||
+      profile.find("technique")->second.size() == 0) {
+    mds.profile["technique"] = "reed_sol_van";
+    pft.profile["technique"] = "reed_sol_van";
+    *ss << "technique not found in profile picking default reed_sol_van" << std::endl;
+  } else {
+    std::string p = profile.find("technique")->second;
+    *ss << "recieved technique as " << p << std::endl;
+    mds.profile["technique"] = p;
+    pft.profile["technique"] = p;
+  }
+  
+  if ((d < k) || (d > k+m-1)) {
+    *ss << "value of d " << d
+        << " must be within [ " << k << "," << k+m-1 << "]" << std::endl;
+    err = -EINVAL;
+    return err;
+  }
+  
+  q = d-k+1;
+  if ((k+m)%q) {
+    nu = q - (k+m)%q;
+    *ss << "Clay: (k+m)%q=" << (k+m)%q
+	<< "q doesn't divide k+m, to use shortening" << std::endl;
+  } else {
+    nu = 0;
+  }
+
+  if (k+m+nu > 254) {
+    err = -EINVAL;
+    return err;
+  }
+  
+  mds.profile["k"] = std::to_string(k+nu);
+  mds.profile["m"] = std::to_string(m);
+  mds.profile["w"] = '8';
+  
+  pft.profile["k"] = '2';
+  pft.profile["m"] = '2';
+  pft.profile["w"] = '8';
+  
+  dout(10) << __func__ << " k:" << k << " m: " << m << " w:" << w << dendl;
+    
+  t = (k+m+nu)/q;
+  sub_chunk_no = pow_int(q, t); 
+
+  return 0;
+}
+
+int ErasureCodeClay::get_repair_sub_chunk_count(const set<int> &want_to_read)
+{
+  int repair_subchunks_count = 1;
+
+  int weight_vector[t];
+  memset(weight_vector, 0, t*sizeof(int));
+
+  for (set<int>::iterator i = want_to_read.begin();
+       i != want_to_read.end(); ++i) {
+    weight_vector[(*i)/q]++;
+  }
+
+  for (int y = 0; y < t; y++) repair_subchunks_count = repair_subchunks_count*(q-weight_vector[y]);
+
+  return sub_chunk_no - repair_subchunks_count;
+}
+
+int ErasureCodeClay::is_repair(const set<int> &want_to_read,
+			       const set<int> &available_chunks) {
+
+  if (includes(
+	       available_chunks.begin(), available_chunks.end(), want_to_read.begin(), want_to_read.end())) return 0;
+
+  if (want_to_read.size() > 1) return 0;
```

Comment:
Excuse me, I have questions that why if want_to_read.size() > 1 then return 0? If so it only repair in single failure case? And if want_to_read.size() = 1 then the code between line 311 and line 321 can be optimized to the following？

int i =*want_to_read.begin();
int lost_node_id = (i < k) ? i: i+nu;
for (int x = 0; x < q; x++) {
&nbsp;&nbsp;&nbsp;&nbsp;int node = (lost_node_id/q)*q+x;
&nbsp;&nbsp;&nbsp;&nbsp;node = (node < k) ? node : node-nu;
&nbsp;&nbsp;&nbsp;&nbsp;if (node != lost_node_id) {//node from same group not erased
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if (available_chunks.find(node) == available_chunks.end()) return 0;//node from same group not available as well
&nbsp;&nbsp;&nbsp;&nbsp;}
}